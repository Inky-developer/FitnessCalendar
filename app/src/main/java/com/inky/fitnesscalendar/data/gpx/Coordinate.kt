package com.inky.fitnesscalendar.data.gpx

import android.location.Location

data class Coordinate(val latitude: Double, val longitude: Double) {
    fun distanceMeters(other: Coordinate, results: FloatArray): Double {
        Location.distanceBetween(
            this.latitude,
            this.longitude,
            other.latitude,
            other.longitude,
            results
        )
        return results[0].toDouble()
    }
}
