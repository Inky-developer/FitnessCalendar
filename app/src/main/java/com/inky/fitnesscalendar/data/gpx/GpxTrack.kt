package com.inky.fitnesscalendar.data.gpx

data class GpxTrack(val name: String, val type: String, val points: List<GpxTrackPoint>) {
    val startTime = points.firstOrNull()?.time
    val endTime = points.lastOrNull()?.time
}
