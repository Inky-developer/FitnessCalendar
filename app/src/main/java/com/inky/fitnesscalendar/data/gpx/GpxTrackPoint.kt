package com.inky.fitnesscalendar.data.gpx

import kotlinx.serialization.KSerializer
import kotlinx.serialization.Serializable
import kotlinx.serialization.descriptors.PrimitiveKind
import kotlinx.serialization.descriptors.PrimitiveSerialDescriptor
import kotlinx.serialization.descriptors.SerialDescriptor
import kotlinx.serialization.encoding.Decoder
import kotlinx.serialization.encoding.Encoder
import java.util.Date

@Serializable
data class GpxTrackPoint(
    val coordinate: Coordinate,
    @Serializable(DateSerializer::class)
    val time: Date
) {
    private object DateSerializer : KSerializer<Date> {
        override val descriptor: SerialDescriptor
            get() = PrimitiveSerialDescriptor("Date", PrimitiveKind.LONG)

        override fun serialize(encoder: Encoder, value: Date) {
            encoder.encodeLong(value.time)
        }

        override fun deserialize(decoder: Decoder) = Date(decoder.decodeLong())
    }
}
