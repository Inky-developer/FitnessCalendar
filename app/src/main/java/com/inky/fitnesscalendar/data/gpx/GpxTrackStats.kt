package com.inky.fitnesscalendar.data.gpx

import com.inky.fitnesscalendar.data.measure.Distance
import com.inky.fitnesscalendar.data.measure.Duration
import com.inky.fitnesscalendar.data.measure.Elevation
import com.inky.fitnesscalendar.data.measure.HeartFrequency
import com.inky.fitnesscalendar.data.measure.Temperature

data class GpxTrackStats(
    val totalDistance: Distance,
    val totalDuration: Duration,
    val movingDuration: Duration,
    val averageHeartFrequency: HeartFrequency?,
    val maxHeartFrequency: HeartFrequency?,
    val averageTemperature: Temperature?,
    val minTemperature: Temperature?,
    val maxTemperature: Temperature?,
    val minHeight: Elevation?,
    val maxHeight: Elevation?
)