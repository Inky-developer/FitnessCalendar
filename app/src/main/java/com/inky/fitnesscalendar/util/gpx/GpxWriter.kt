package com.inky.fitnesscalendar.util.gpx

import android.content.Context
import android.util.Xml
import com.inky.fitnesscalendar.R
import com.inky.fitnesscalendar.db.entities.RichActivity
import com.inky.fitnesscalendar.db.entities.Track
import org.xmlpull.v1.XmlSerializer
import java.io.Writer
import java.time.ZoneOffset
import java.time.format.DateTimeFormatter

class GpxWriter {
    companion object {
        fun write(activity: RichActivity, track: Track, context: Context, writer: Writer) {
            val serializer = Serializer(writer)
            return serializer.document {
                tag("gpx") {
                    attribute("creator", context.getString(R.string.app_name))
                    attribute("version", "1.1")
                    attribute("xmlns", "http://www.topografix.com/GPX/1/1")

                    tag("trk") {
                        writeTrack(activity, track)
                    }
                }
            }
        }
    }
}

private fun TagScope.writeTrack(activity: RichActivity, track: Track) {
    tag("name") { text(activity.type.name) }
    tag("desc") { text(activity.activity.description) }
    tag("type") { text(activity.type.name) }
    tag("trkseg") {
        for (point in track.points) {
            tag("trkpt") {
                attribute("lat", point.coordinate.latitude.toString())
                attribute("lon", point.coordinate.longitude.toString())

                tag("time") {
                    text(
                        point.time.toInstant().atOffset(ZoneOffset.UTC)
                            .format(DateTimeFormatter.ISO_OFFSET_DATE_TIME)
                    )
                }

                if (point.elevation != null) {
                    tag("ele") { text(point.elevation.meters.toString()) }
                }

                // TODO: Include other attributes like temperature and heart rate
            }
        }
    }
}

private class Serializer(writer: Writer) {
    private val serializer: XmlSerializer = Xml.newSerializer().apply {
        setOutput(writer)
    }

    inline fun document(inner: DocumentScope.() -> Unit) {
        val scope = DocumentScope(serializer)
        serializer.startDocument("UTF-8", true)
        inner(scope)
        serializer.endDocument()
    }
}

private open class DocumentScope(val serializer: XmlSerializer) {
    inline fun tag(name: String, inner: TagScope.() -> Unit) {
        val scope = TagScope(serializer)
        serializer.startTag(null, name)
        inner(scope)
        serializer.endTag(null, name)
    }
}

private class TagScope(serializer: XmlSerializer) : DocumentScope(serializer) {
    fun attribute(name: String, value: String) {
        serializer.attribute(null, name, value)
    }

    fun text(text: String) {
        serializer.text(text)
    }
}