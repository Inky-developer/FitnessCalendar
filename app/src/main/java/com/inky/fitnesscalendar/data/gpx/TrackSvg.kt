package com.inky.fitnesscalendar.data.gpx

import com.inky.fitnesscalendar.db.entities.CoordinateRect
import com.inky.fitnesscalendar.db.entities.CoordinateRect.Companion.calculateBounds
import com.inky.fitnesscalendar.util.gpx.simplify
import kotlinx.serialization.Serializable


@Serializable
data class TrackSvg(val points: List<Coordinate>, val bounds: CoordinateRect) {
    companion object {
        fun fromPoints(points: List<Coordinate>): TrackSvg? {
            val bounds = points.calculateBounds() ?: return null
            return TrackSvg(points, bounds)
        }

        fun GpxTrack.toTrackSvg() = fromPoints(simplify(points.map { it.coordinate }))
    }
}