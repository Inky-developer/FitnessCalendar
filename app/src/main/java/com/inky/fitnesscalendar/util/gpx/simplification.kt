package com.inky.fitnesscalendar.util.gpx

import com.inky.fitnesscalendar.data.gpx.Coordinate
import java.util.PriorityQueue
import kotlin.math.absoluteValue

private const val TAG = "util/gpx/simplification"

/**
 * Simplifies the track by removing the most irrelevant points of the track until only a fixed number
 * of points is left.
 * A point is irrelevant if the triangle formed by its predecessor and ancestor has a low area.
 *
 * @param maxNumPoints Keep at most that many points
 */
fun simplify(points: List<Coordinate>, maxNumPoints: Int = 100): List<Coordinate> {
    val numPointsToRemove = points.size - maxNumPoints
    if (numPointsToRemove <= 0) {
        return points
    }

    val triangles = points.indices.map { index ->
        Triangle(
            area = calculateArea(
                points.getOrNull(index - 1),
                points[index],
                points.getOrNull(index + 1)
            ),
            coordinate = points[index],
            predecessorIndex = index - 1,
            index = index,
            successorIndex = index + 1,
        )
    }
    // TODO: This Priority Queue is pretty useless, since `remove` is O(n)
    val sortedTriangles = PriorityQueue(triangles)
    for (i in 0..<numPointsToRemove) {
        val toRemove = sortedTriangles.first()
        sortedTriangles.remove(toRemove)

        triangles.getOrNull(toRemove.predecessorIndex)?.apply {
            sortedTriangles.remove(this)
            successorIndex = toRemove.successorIndex
            updateArea(triangles, points)
            sortedTriangles.add(this)
        }
        triangles.getOrNull(toRemove.successorIndex)?.apply {
            sortedTriangles.remove(this)
            predecessorIndex = toRemove.predecessorIndex
            updateArea(triangles, points)
            sortedTriangles.add(this)
        }
    }

    val pointsToKeep = sortedTriangles.map { it.index }.toSet()
    val coordinates = points
        .asSequence()
        .filterIndexed { index, _ -> pointsToKeep.contains(index) }
        .toList()
    return coordinates
}

// Lets hope this is correct
private fun calculateArea(a: Coordinate?, b: Coordinate, c: Coordinate?): Float {
    // The start and end point should never get removed and thus will be assigned infinite area
    if (a == null || c == null) return Float.POSITIVE_INFINITY
    return 0.5f * (a.latitude * (b.longitude - c.longitude) + b.latitude * (c.longitude - a.longitude) + c.latitude * (a.longitude - b.longitude)).absoluteValue.toFloat()
}

private data class Triangle(
    var area: Float,
    val coordinate: Coordinate,
    var predecessorIndex: Int,
    val index: Int,
    var successorIndex: Int
) : Comparable<Triangle> {
    fun updateArea(triangles: List<Triangle>, points: List<Coordinate>) {
        val a = triangles.getOrNull(predecessorIndex)?.coordinate
        val c = triangles.getOrNull(successorIndex)?.coordinate
        area = calculateArea(a, points[index], c)
    }

    override fun compareTo(other: Triangle): Int {
        return area.compareTo(other.area)
    }
}