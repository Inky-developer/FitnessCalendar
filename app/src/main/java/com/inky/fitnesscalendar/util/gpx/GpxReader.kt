package com.inky.fitnesscalendar.util.gpx

import android.util.Log
import android.util.Xml
import com.inky.fitnesscalendar.data.gpx.Coordinate
import com.inky.fitnesscalendar.data.gpx.GpxTrack
import com.inky.fitnesscalendar.data.gpx.GpxTrackPoint
import com.inky.fitnesscalendar.data.measure.Elevation
import com.inky.fitnesscalendar.data.measure.HeartFrequency
import com.inky.fitnesscalendar.data.measure.Temperature
import org.xmlpull.v1.XmlPullParser
import org.xmlpull.v1.XmlPullParserException
import java.io.InputStream
import java.time.Instant
import java.time.format.DateTimeParseException
import java.util.Date


private const val TAG = "GpxReader"

private const val TRACK_POINT_EXTENSIONS_GARMIN = "ns3:TrackPointExtension"
private const val EXTENSION_GARMIN_TEMPERATURE = "ns3:atemp"
private const val EXTENSION_GARMIN_HEART_RATE = "ns3:hr"

class GpxReader(val tracks: List<GpxTrack>) {
    companion object {
        fun read(inputStream: InputStream): GpxReader? {
            val parser: XmlPullParser = Xml.newPullParser()
            parser.setFeature(XmlPullParser.FEATURE_PROCESS_NAMESPACES, false)

            try {
                parser.setInput(inputStream, null)
                parser.nextTag()
                return readFeed(parser)
            } catch (e: XmlPullParserException) {
                Log.e(TAG, "Could not parse feed: $e")
                return null
            }
        }

        @Throws(XmlPullParserException::class)
        private fun readFeed(parser: XmlPullParser): GpxReader {
            val tracks = readTag(parser, "gpx") {
                when (it.name) {
                    "trk" -> readTrack(it)

                    else -> {
                        skipTag(it)
                        null
                    }
                }
            }

            if (tracks.isEmpty()) {
                throw XmlPullParserException("Gpx file contained no track")
            }
            return GpxReader(tracks)
        }

        @Throws(XmlPullParserException::class)
        private fun readTrack(parser: XmlPullParser): GpxTrack {
            var trackName = ""
            var trackDescription = ""
            var trackType = ""

            val trackPoints = readTag(parser, "trk") {
                when (it.name) {
                    "name" -> {
                        trackName = readText(it, "name")
                        null
                    }

                    "desc" -> {
                        trackDescription = readText(it, "desc")
                        null
                    }

                    "type" -> {
                        trackType = readText(it, "type")
                        null
                    }

                    "trkseg" -> readTrackSegment(it)

                    else -> {
                        skipTag(it)
                        null
                    }
                }
            }

            if (trackDescription.isNotBlank()) {
                trackName += "\n\n" + trackDescription
            }
            return GpxTrack(name = trackName, type = trackType, points = trackPoints.flatten())
        }

        @Throws(XmlPullParserException::class)
        private fun readTrackSegment(parser: XmlPullParser): List<GpxTrackPoint> {
            return readTag(parser, "trkseg") {
                when (it.name) {
                    "trkpt" -> {
                        val lat = it.getAttributeValue(null, "lat")
                        val long = it.getAttributeValue(null, "lon")
                        val coordinate = try {
                            Coordinate(lat.toDouble(), long.toDouble())
                        } catch (e: NumberFormatException) {
                            throw XmlPullParserException("Cannot parse coordinate (lat=$lat, lon=$long)")
                        }

                        val attributes = TrackPointAttributes()
                        readTag(it, "trkpt") { inner ->
                            when (inner.name) {
                                "time" -> {
                                    if (attributes.time != null) {
                                        throw XmlPullParserException("Track point contains multiple times")
                                    }
                                    attributes.time = readTime(inner)
                                }

                                "ele" -> {
                                    if (attributes.elevation != null) {
                                        throw XmlPullParserException("Track point contains multiple elevations")
                                    }
                                    attributes.elevation = readElevation(inner)
                                }

                                "extensions" -> {
                                    readTrackPointExtensions(parser, attributes)
                                }

                                else -> skipTag(inner)
                            }
                        }

                        GpxTrackPoint(
                            coordinate = coordinate,
                            time = attributes.time
                                ?: throw XmlPullParserException("Track point without required tag time"),
                            elevation = attributes.elevation,
                            heartFrequency = attributes.heartRate,
                            temperature = attributes.temperature
                        )
                    }

                    else -> {
                        skipTag(it)
                        null
                    }
                }
            }
        }

        @Throws(XmlPullParserException::class)
        private fun readTrackPointExtensions(
            parser: XmlPullParser,
            attributes: TrackPointAttributes
        ) {
            readTag(parser, "extensions") {
                when (it.name) {
                    TRACK_POINT_EXTENSIONS_GARMIN ->
                        readTrackPointExtensionsGarmin(parser, attributes)

                    else -> skipTag(parser)
                }
            }

        }

        @Throws(XmlPullParserException::class)
        private fun readTrackPointExtensionsGarmin(
            parser: XmlPullParser,
            attributes: TrackPointAttributes
        ) {
            readTag(parser, TRACK_POINT_EXTENSIONS_GARMIN) {
                when (it.name) {
                    EXTENSION_GARMIN_TEMPERATURE -> {
                        val temperatureString = readText(parser, EXTENSION_GARMIN_TEMPERATURE)
                        attributes.temperature = Temperature(
                            celsius = temperatureString.toFloatOrNull()
                                ?: throw XmlPullParserException("Invalid temperature: $temperatureString")
                        )
                    }

                    EXTENSION_GARMIN_HEART_RATE -> {
                        val heartRateString = readText(parser, EXTENSION_GARMIN_HEART_RATE)
                        attributes.heartRate = HeartFrequency(
                            bpm = heartRateString.toFloatOrNull()
                                ?: throw XmlPullParserException("Invalid heart rate: $heartRateString")
                        )
                    }

                    else -> skipTag(it)
                }
            }
        }

        @Throws(XmlPullParserException::class)
        private fun readTime(parser: XmlPullParser): Date {
            val dateString = readText(parser, "time")
            try {
                return Date.from(Instant.parse(dateString))
            } catch (e: DateTimeParseException) {
                throw XmlPullParserException("Invalid date string: $dateString")
            }
        }

        @Throws(XmlPullParserException::class)
        private fun readElevation(parser: XmlPullParser): Elevation {
            val elevationString = readText(parser, "ele")
            val elevation = elevationString.toFloatOrNull()
                ?: throw XmlPullParserException("Invalid elevation: $elevationString")
            return Elevation(meters = elevation)
        }

        @Throws(XmlPullParserException::class)
        private inline fun <T : Any> readTag(
            parser: XmlPullParser,
            tagName: String,
            handler: (XmlPullParser) -> T?
        ): List<T> {
            val result = mutableListOf<T>()

            parser.require(XmlPullParser.START_TAG, null, tagName)
            while (parser.next() != XmlPullParser.END_TAG) {
                if (parser.eventType != XmlPullParser.START_TAG) {
                    continue
                }

                val handlerResult = handler(parser)
                if (handlerResult != null) {
                    result.add(handlerResult)
                }
            }

            return result
        }

        @Throws(XmlPullParserException::class)
        private fun readText(parser: XmlPullParser, tag: String): String {
            parser.require(XmlPullParser.START_TAG, null, tag)
            val text = if (parser.next() == XmlPullParser.TEXT) {
                val result = parser.text
                parser.nextTag()
                result
            } else {
                ""
            }
            parser.require(XmlPullParser.END_TAG, null, tag)
            return text
        }

        @Throws(XmlPullParserException::class)
        private fun skipTag(parser: XmlPullParser) {
            if (parser.eventType != XmlPullParser.START_TAG) {
                throw XmlPullParserException("Must be at start tag")
            }
            var depth = 1
            while (depth != 0) {
                when (parser.next()) {
                    XmlPullParser.END_TAG -> depth -= 1
                    XmlPullParser.START_TAG -> depth += 1
                }
            }
        }

    }

    private data class TrackPointAttributes(
        var elevation: Elevation? = null,
        var time: Date? = null,
        var temperature: Temperature? = null,
        var heartRate: HeartFrequency? = null
    )

}