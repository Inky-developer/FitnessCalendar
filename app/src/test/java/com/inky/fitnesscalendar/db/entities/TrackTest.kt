package com.inky.fitnesscalendar.db.entities

import com.inky.fitnesscalendar.data.gpx.Coordinate
import com.inky.fitnesscalendar.util.gpx.simplify
import org.junit.Assert.assertEquals
import org.junit.Test

class TrackTest {

    @Test
    fun testSimplify() {
        val trackPoints = listOf(
            Coordinate(0.0, 0.0),
            Coordinate(1.0, 1.0),
            Coordinate(2.0, -1.0),
            Coordinate(3.0, -1.5)
        )
        val simplifiedTrack = simplify(trackPoints, maxNumPoints = 3)

        assertEquals(
            listOf(
                Coordinate(0.0, 0.0),
                Coordinate(1.0, 1.0),
                Coordinate(3.0, -1.5)
            ),
            simplifiedTrack,
        )
    }
}