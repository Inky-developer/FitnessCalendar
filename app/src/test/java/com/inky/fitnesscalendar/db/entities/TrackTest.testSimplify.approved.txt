TrackSimplification[0] = Coordinate(latitude=0.0, longitude=0.0)
TrackSimplification[1] = Coordinate(latitude=1.0, longitude=1.0)
TrackSimplification[2] = Coordinate(latitude=3.0, longitude=-1.5)
